Userland/Libraries/LibWebSocket/WebSocket.cpp

```diff
@@ -504,16 +499,34 @@ void WebSocket::read_frame()
         // We can safely ignore the pong
         return;
     }
-    if (op_code == WebSocket::OpCode::Continuation) {
-        // FIXME: Support fragmented frames
-        TODO();
+    if (!is_final_frame) {
+        if (op_code != WebSocket::OpCode::Continuation) {
+            // First fragmented message
+            m_initial_fragment_opcode = op_code;
+        }
+        // First and next fragmented message
+        m_fragmented_data_buffer.append(payload.data(), payload_length);
+        return;
+    }
+    auto send_last_frame = is_final_frame && op_code == WebSocket::OpCode::Continuation;
+    if (send_last_frame) {
+        // Last fragmented message
+        m_fragmented_data_buffer.append(payload.data(), payload_length);
+        op_code = m_initial_fragment_opcode;
+        payload = m_fragmented_data_buffer;
     }
     if (op_code == WebSocket::OpCode::Text) {
         notify_message(Message(payload, true));
+        if (send_last_frame) {
```

Comment:
er wait. I think you wanted to write `is_final_frame` here? `send_last_frame` requires the op_code to be `Continuation`

Reply:
but we already know here that `is_final_frame` is true, because we returned above if it was false. There's some logic simplification that could go on here I think

Reply:
Apologies, I think the var name that I used is unclear.
My intention was to clear the `m_fragmented_buffer` once we need to send the fragmented message as a whole. Otherwise as far as I understand I think I would hold on to the buffer in a scenario where we:
 1. Receive fragmented message of let's say 4 MB
 2. Then receive normal messages
 3. Then that 4 MB is still in the memory.

So when I want to clear the `m_fragmented_buffer` I cannot rely on `is_final_frame` bit for that, because `is_final_frame` might be sent for messages that are not fragmented, meaning we don't use the `m_fragmented_buffer` in these scenarios at all.

A better name would be: 

```cpp
auto is_last_fragmented_frame = is_final_frame && op_code == WebSocket::OpCode::Continuation;
```

We need to clear the `m_fragmented_buffer` because otherwise WebSocket could break in a situation where:
 - We receive a fragmented message
 - Then receive a normal message in-between
 - Then receive fragmented message again

I have updated the PR with a little bit more cleaner code, but it is not really memory efficient.

I have retested the code and it work, but there seem to be a bug that seem to be unrelated to `LibWebSocket`. In some situations (quick refreshes in Ladybird) it seems like the TCP Socket itself gets stuck awaiting to fetch some bytes.

Reply:
Much cleaner now, thanks! I wouldn't worry about the memory efficiency too much until we start seeing it show up in profiles